standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java

```diff
@@ -2225,45 +2238,68 @@ public AggrStats get_aggr_stats_for(String catName, String dbName, String tblNam
     }
 
     List<String> allPartNames = rawStore.listPartitionNames(catName, dbName, tblName, (short) -1);
+    StatsType type = StatsType.PARTIAL;
     if (partNames.size() == allPartNames.size()) {
       colStats = sharedCache.getAggrStatsFromCache(catName, dbName, tblName, colNames, StatsType.ALL);
       if (colStats != null) {
         return new AggrStats(colStats, partNames.size());
       }
+      type = StatsType.ALL;
     } else if (partNames.size() == (allPartNames.size() - 1)) {
       String defaultPartitionName = MetastoreConf.getVar(getConf(), ConfVars.DEFAULTPARTITIONNAME);
       if (!partNames.contains(defaultPartitionName)) {
-        colStats =
-            sharedCache.getAggrStatsFromCache(catName, dbName, tblName, colNames, StatsType.ALLBUTDEFAULT);
+        colStats = sharedCache.getAggrStatsFromCache(catName, dbName, tblName, colNames, StatsType.ALLBUTDEFAULT);
         if (colStats != null) {
           return new AggrStats(colStats, partNames.size());
         }
+        type = StatsType.ALLBUTDEFAULT;
       }
     }
+
     LOG.debug("Didn't find aggr stats in cache. Merging them. tblName= {}, parts= {}, cols= {}",
         tblName, partNames, colNames);
-    MergedColumnStatsForPartitions mergedColStats =
-        mergeColStatsForPartitions(catName, dbName, tblName, partNames, colNames, sharedCache);
+    MergedColumnStatsForPartitions mergedColStats = mergeColStatsForPartitions(catName, dbName, tblName,
+              partNames, colNames, sharedCache, type, writeIdList);
+    if (mergedColStats == null) {
+      LOG.info("Aggregate stats of partition " + TableName.getQualified(catName, dbName, tblName) + "." +
+              partNames + " for columns " + colNames + " is not present in cache. Getting it from raw store");
+      return rawStore.get_aggr_stats_for(catName, dbName, tblName, partNames, colNames, writeIdList);
+    }
     return new AggrStats(mergedColStats.getColStats(), mergedColStats.getPartsFound());
   }
 
   private MergedColumnStatsForPartitions mergeColStatsForPartitions(
       String catName, String dbName, String tblName, List<String> partNames, List<String> colNames,
-      SharedCache sharedCache) throws MetaException {
+      SharedCache sharedCache, StatsType type, String writeIdList) throws MetaException {
     final boolean useDensityFunctionForNDVEstimation =
         MetastoreConf.getBoolVar(getConf(), ConfVars.STATS_NDV_DENSITY_FUNCTION);
     final double ndvTuner = MetastoreConf.getDoubleVar(getConf(), ConfVars.STATS_NDV_TUNER);
     Map<ColumnStatsAggregator, List<ColStatsObjWithSourceInfo>> colStatsMap = new HashMap<>();
-    boolean areAllPartsFound = true;
-    long partsFound = 0;
+    long partsFound = partNames.size();
+    Map<List<String>, Long> partNameToWriteId = writeIdList != null ? new HashMap<>() : null;
```

Comment:
Is the logic for get_aggr_stats_for different than ObjectStore? I see in ObjectStore, stats are aggregated anyway as long as partitions are valid:
```
    for (Partition part : parts) {
        if (!isCurrentStatsValidForTheQuery(part, part.getWriteId(), writeIdList, false)) {
          .......
          return null;
        }
      }
    }
    return get_aggr_stats_for(catName, dbName, tblName, partNames, colNames);
```

Reply:
The behavior is same as objet store ...if partition is not present then ignore, but if partition is present and is not compatible then return null.